src/run.c

```diff
@@ -276,52 +308,100 @@ Port io_open_file(Net* net, Book* book, Port argm) {
     }
   }
 
-  fprintf(stderr, "io_open_file: too many open files\n");
-
+  fprintf(stderr, "io_open: too many open files\n");
   return new_port(ERA, 0);
 }
 
 // Closes a file, reclaiming the file descriptor.
-Port io_close_file(Net* net, Book* book, Port argm) {
-  FILE* fp = readback_file(peek(net, argm));
+Port io_close(Net* net, Book* book, Port argm) {
+  FILE* fp = readback_file(argm);
   if (fp == NULL) {
-    fprintf(stderr, "io_close_file: failed to close\n");
+    fprintf(stderr, "io_close: failed to close\n");
     return new_port(ERA, 0);
   }
 
   int err = fclose(fp) != 0;
   if (err != 0) {
-    fprintf(stderr, "io_close_file: failed to close: %i\n", err);
+    fprintf(stderr, "io_close: failed to close: %i\n", err);
     return new_port(ERA, 0);
   }
 
   FILE_POINTERS[get_u24(get_val(argm))] = NULL;
-
   return new_port(ERA, 0);
 }
 
-// Writes a string to a file.
+// Writes a list of bytes to a file.
 // `argm` is a tuple (CON node) of the
-// file descriptor and string to write.
+// file descriptor and list of bytes to write.
 Port io_write(Net* net, Book* book, Port argm) {
-  if (get_tag(peek(net, argm)) != CON) {
-    fprintf(stderr, "io_write: expected tuple, but got %u\n", get_tag(peek(net, argm)));
+  if (get_tag(argm) != CON) {
+    fprintf(stderr, "io_write: expected tuple, but got %u\n", get_tag(argm));
     return new_port(ERA, 0);
   }
 
   Pair args = node_load(net, get_val(argm));
-  FILE* fp = readback_file(peek(net, get_fst(args)));
-  Str str = readback_str(net, book, get_snd(args));
+  FILE* fp = readback_file(expand(net, book, get_fst(args)));
+  Bytes bytes = readback_bytes(net, book, get_snd(args));
 
   if (fp == NULL) {
     fprintf(stderr, "io_write: invalid file descriptor\n");
+    free(bytes.buf);
     return new_port(ERA, 0);
   }
 
-  if (fputs(str.text_buf, fp) == EOF) {
+  if (fwrite(bytes.buf, sizeof(char), bytes.len, fp) != bytes.len) {
     fprintf(stderr, "io_write: failed to write\n");
   }
 
+  free(bytes.buf);
+  return new_port(ERA, 0);
+}
+
+// Seeks to a position in a file.
+// `argm` is a 3-tuple (CON fd (CON offset whence)), where
+// - fd is a file descriptor
+// - offset is a signed byte offset
+// - whence is what that offset is relative to:
+//    - 0 (SEEK_SET): beginning of file
+//    - 1 (SEEK_CUR): current position of the file pointer
+//    - 2 (SEEK_END): end of the file
+Port io_seek(Net* net, Book* book, Port argm) {
+  if (get_tag(argm) != CON) {
+    fprintf(stderr, "io_seek: expected first tuple, but got %u\n", get_tag(argm));
+    return new_port(ERA, 0);
+  }
+
+  Pair args1 = node_load(net, get_val(argm));
```

Comment:
Every time we read an argument node we want to call expand no?
So here `(IO/call "SEEK" SeekArgsRef)` where SeekArgsRef is not a variable or a tuple, but a reference to a definition, we'll still get the error no?

In `readback_ctr` this is already being done for example

Reply:
Also, if we want external people to be able to write these functions, we should probably encapsulate this with some nicer api for reading arguments from a tuple, just like we do for bytes and strings

Reply:
The `argm` passed to io functions is parsed through `readback_ctr`, so it is already expanded. I've added a `readback_tup`, I'll expose this in another PR.